File: ios/Camera Kit/CameraKitViewManager.m
//
//  CameraKitViewManager.m
//  react-native-snapchat-camera-kit
//
//  Created by Rıdvan Altun on 25.03.2023.
//

#import <Foundation/Foundation.h>

#import <React/RCTViewManager.h>

@interface RCT_EXTERN_REMAP_MODULE(CameraKitView, CameraKitViewManager, RCTViewManager)

- (dispatch_queue_t)methodQueue
{
  return dispatch_get_main_queue();
}

+ (BOOL)requiresMainQueueSetup
{
  return YES;
}

// Camera View Properties
RCT_EXPORT_VIEW_PROPERTY(isActive, BOOL);
RCT_EXPORT_VIEW_PROPERTY(preset, NSString);
RCT_EXPORT_VIEW_PROPERTY(initialLens, NSDictionary);
RCT_EXPORT_VIEW_PROPERTY(initialCameraFacing, NSString);
RCT_EXPORT_VIEW_PROPERTY(lensGroups, NSArray);
RCT_EXPORT_VIEW_PROPERTY(zoom, BOOL);
RCT_EXPORT_VIEW_PROPERTY(focus, BOOL);
RCT_EXPORT_VIEW_PROPERTY(torch, NSDictionary);

// Camera View Events
RCT_EXPORT_VIEW_PROPERTY(onInitialized, RCTDirectEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onLensChanged, RCTDirectEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onPhotoTaken, RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onVideoRecordingFinished, RCTBubblingEventBlock);
RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock);

// Camera View Functions - (Recording)
RCT_EXTERN_METHOD(startRecording:(nonnull NSNumber *)node options:(NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(pauseRecording:(nonnull NSNumber *)node resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(resumeRecording:(nonnull NSNumber *)node resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(cancelRecording:(nonnull NSNumber *)node resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(stopRecording:(nonnull NSNumber *)node resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);

// Camera View Functions - (Miscellaneous)
RCT_EXTERN_METHOD(takePhoto:(nonnull NSNumber *)node options:(NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(flipCamera:(nonnull NSNumber *)node resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(focus:(nonnull NSNumber *)node x:(nonnull NSNumber *)x y:(nonnull NSNumber *)y resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(zoom:(nonnull NSNumber *)node factor:(nonnull NSNumber *)factor resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(clearLenses:(nonnull NSNumber *)node onClearLensCallback:(RCTResponseSenderBlock)onClearLensCallback);
RCT_EXTERN_METHOD(adjustLensesVolume:(nonnull NSNumber *)node mute:(BOOL)mute onAdjustLensesVolumeCallback:(RCTResponseSenderBlock)onAdjustLensesVolumeCallback);
RCT_EXTERN_METHOD(getLensesByGroupId:(nonnull NSNumber *)node lensGroups:(NSArray *)lensGroups onGetLensesByGroupIdCallback:(RCTResponseSenderBlock)onGetLensesByGroupIdCallback);
RCT_EXTERN_METHOD(changeLensById:(nonnull NSNumber *)node lensId:(NSString *)lensId lensGroups:(NSArray *)lensGroups launchDataMap:(NSDictionary *)launchDataMap resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(checkBlurSupporting:(nonnull NSNumber *)node onCheckBlurSupportingCallback:(RCTResponseSenderBlock)onCheckBlurSupportingCallback);
RCT_EXTERN_METHOD(checkToneModeSupporting:(nonnull NSNumber *)node onCheckToneModeSupportingCallback:(RCTResponseSenderBlock)onCheckToneModeSupportingCallback);
RCT_EXTERN_METHOD(adjustBlur:(nonnull NSNumber *)node amount:(nonnull NSNumber *)amount);
RCT_EXTERN_METHOD(adjustTone:(nonnull NSNumber *)node amount:(nonnull NSNumber *)amount);

// Miscellaneous Functions
_RCT_EXTERN_REMAP_METHOD(init, initMethod:(NSDictionary *)options, false);
RCT_EXTERN_METHOD(getMetadata:(RCTResponseSenderBlock)onGetMetadataCallback);

@end

File: ios/Camera Kit/CameraKitViewManager.swift
//
//  CameraKitViewManager.swift
//  DemoApp
//
//  Created by Rıdvan Altun on 25.03.2023.
//

import Foundation
import SCSDKCameraKit

@objc(CameraKitViewManager)
final class CameraKitViewManager: RCTViewManager {
    
    var apiKey: String? = nil
    var applicationId: String? = nil
    
    override final func view() -> UIView! {
        return CameraKitView(apiKey: apiKey, applicationId: applicationId)
    }
    
    // Camera View Functions - (Recording)
    
    @objc
    final func startRecording(_ node: NSNumber, options: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.startRecording(options: options, promise: promise)
    }
    
    @objc
    final func pauseRecording(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.pauseRecording(promise: promise)
    }
    
    @objc
    final func resumeRecording(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.resumeRecording(promise: promise)
    }
    
    @objc
    final func cancelRecording(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.cancelRecording(promise: promise)
    }
    
    @objc
    final func stopRecording(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.stopRecording(promise: promise)
    }
    
    // Camera View Functions - (Miscellaneous)
    
    @objc
    final func takePhoto(_ node: NSNumber, options: NSDictionary, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.takePhoto(options: options, promise: promise)
    }
    
    @objc
    final func flipCamera(_ node: NSNumber, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.flipCamera(promise: promise)
    }
    
    @objc
    final func zoom(_ node: NSNumber, factor: Double, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.zoomMethod(factor: factor, promise: promise)
    }
    
    @objc
    final func focus(_ node: NSNumber, x: Double, y: Double, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        component.focusMethod(point: CGPoint(x: x, y: y), promise: promise)
    }
    
    @objc
    final func clearLenses(_ node: NSNumber, onClearLensCallback: @escaping RCTResponseSenderBlock) {
        let component = getCameraKitView(withTag: node)
        component.clearLenses(callback: onClearLensCallback)
    }
    
    @objc
    final func adjustLensesVolume(_ node: NSNumber, mute: Bool, onAdjustLensesVolumeCallback: @escaping RCTResponseSenderBlock) {
        let component = getCameraKitView(withTag: node)
        component.adjustLensesVolume(mute: mute, callback: onAdjustLensesVolumeCallback)
    }
    
    @objc
    final func getLensesByGroupId(_ node: NSNumber, lensGroups: NSArray, onGetLensesByGroupIdCallback: @escaping RCTResponseSenderBlock) {
        let component = getCameraKitView(withTag: node)
        component.getLensesByGroupId(lensGroups: lensGroups, callback: onGetLensesByGroupIdCallback)
    }
    
    @objc
    final func changeLensById(_ node: NSNumber, lensId: NSString, lensGroups: NSArray, launchDataMap: NSDictionary?, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        let promise = Promise(resolver: resolve, rejecter: reject)
        let component = getCameraKitView(withTag: node)
        
        component.applyLensById(
            lensId: lensId,
            lensGroups: lensGroups,
            launchDataMap: launchDataMap,
            promise: promise
        )
    }
    
    @objc
    final func checkBlurSupporting(_ node: NSNumber, onCheckBlurSupportingCallback: @escaping RCTResponseSenderBlock) {
        let component = getCameraKitView(withTag: node)
        component.checkBlurSupporting(callback: onCheckBlurSupportingCallback)
    }
    
    @objc
    final func checkToneModeSupporting(_ node: NSNumber, onCheckToneModeSupportingCallback: @escaping RCTResponseSenderBlock) {
        let component = getCameraKitView(withTag: node)
        component.checkToneModeSupporting(callback: onCheckToneModeSupportingCallback)
    }
    
    @objc
    final func adjustBlur(_ node: NSNumber, amount: NSNumber) {
        let component = getCameraKitView(withTag: node)
        let doubleAmount = Double(truncating: amount)
        component.adjustBlur(amount: doubleAmount)
    }
    
    @objc
    final func adjustTone(_ node: NSNumber, amount: NSNumber) {
        let component = getCameraKitView(withTag: node)
        let doubleAmount = Double(truncating: amount)
        component.adjustTone(amount: doubleAmount)
    }
    
    // Miscellaneous Functions
    
    @objc
    final func initMethod(_ options: NSDictionary) {
        self.apiKey = options.value(forKey: "apiKey") as? String
        self.applicationId = options.value(forKey: "applicationId") as? String
    }
    
    @objc
    final func getMetadata(_ onGetMetadataCallback: RCTResponseSenderBlock) {
        let err: NSNull = NSNull()
        
        let supported = CameraKitHelper.supported()
        
        let meta = MetaModel(supported: supported).toBridge()
        
        onGetMetadataCallback([meta, err])
    }
    
    private func getCameraKitView(withTag tag: NSNumber) -> CameraKitView {
        // swiftlint:disable force_cast
        return bridge.uiManager.view(forReactTag: tag) as! CameraKitView
    }
}
